packages/contracts/src/VaultPlugin.sol

```diff
@@ -0,0 +1,157 @@
+        transfersEnabled = !transfersEnabled;
+    }
+
+    /// @notice Returns the total assets held by the DAO
+    /// @dev See {IERC4626-totalAssets}.
+    function totalAssets() public view virtual override returns (uint256) {
+        return IERC20MetadataUpgradeable(asset()).balanceOf(address(dao()));
+    }
+
+    /// @notice Overrides the _deposit function from the ERC4626 contract to deposit assets into the DAO instead of this
+    /// contract.
+    /// @dev See {IERC4626-_deposit}.
+    /// @param caller The address of the caller.
+    /// @param receiver The address of the receiver.
+    /// @param assets The amount of assets to deposit.
+    /// @param shares The amount of shares to issue.
+    function _deposit(address caller, address receiver, uint256 assets, uint256 shares) internal virtual override {
+        // If _asset is ERC777, `transferFrom` can trigger a reenterancy BEFORE the transfer happens through the
+        // `tokensToSend` hook. On the other hand, the `tokenReceived` hook, that is triggered after the transfer,
+        // calls the vault, which is assumed not malicious.
+        //
+        // Conclusion: we need to do the transfer before we mint so that any reentrancy would happen before the
+        // assets are transfered and before the shares are minted, which is a valid state.
+        // slither-disable-next-line reentrancy-no-eth
+        SafeERC20Upgradeable.safeTransferFrom({
+            token: IERC20Upgradeable(asset()),
+            from: caller,
+            to: address(dao()),
+            value: assets
+        });
+        _mint(receiver, shares);
+
+        emit Deposit(caller, receiver, assets, shares);
+    }
+
+    /// @notice Overrides the _withdraw function from the ERC4626 contract to withdraw assets from the DAO instead of
+    /// this contract.
+    /// @dev See {IERC4626-_withdraw}.
+    /// @param caller The address of the caller.
+    /// @param receiver The address of the receiver.
+    /// @param owner The address of the owner.
+    /// @param assets The amount of assets to withdraw.
+    /// @param shares The amount of shares to burn.
+    function _withdraw(
+        address caller,
+        address receiver,
+        address owner,
+        uint256 assets,
+        uint256 shares
+    )
+        internal
+        virtual
+        override
+    {
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // If _asset is ERC777, `transfer` can trigger a reentrancy AFTER the transfer happens through the
+        // `tokensReceived` hook. On the other hand, the `tokensToSend` hook, that is triggered before the transfer,
+        // calls the vault, which is assumed not malicious.
+        //
+        // Conclusion: we need to do the transfer after the burn so that any reentrancy would happen after the
+        // shares are burned and after the assets are transfered, which is a valid state.
+        _burn(owner, shares);
```

Comment:
I think I need a bit of clarity on how we are going to calculate the burn function in associate with withdrawing the underlining asset from the DAO.

From my understanding, a user should be able to withdraw a portion in relation to the initial amount the deposit to the DAO. Since depositing with this plugin is not the only way a DAO can increase the asset help by the DAO treasury, I fail to understand how this will function.